### ISSUE TITLE
paket install is ignoring paket.local file. paket restore works fine

### ISSUE THREAD

**Issue by creib** [fsprojects/Paket#1925]
### Description

When updating a project using paket install, the .local file is ignored.
When updating a project using paket restore it works fine.
### Expected behavior

When updating a project using paket install, the versions from the paket.local file should be used if specified.

**Comment by theimowski** [comment]
This is by design - paket.local is taken into account only upon restore. If it influenced restore, one would end up with paket.lock containg a local version which is what we want to prevent

**Comment by 0x53A** [comment]
Even if it is by design, I think it is totally unintuitive.

In my (not that important) opinion, a restore after an install should be a NOP, but if I understand it correctly, the install would put the original version in the packages directory, and the restore would then overwrite it with the local version?

Wouldn't it be possible to calculate the paket.lock from the version(s) on nuget, but then immediately afterwards parse the paket.local and copy over the local files?

**Comment by theimowski** [comment]
> the install would put the original version in the packages directory, and the restore would then overwrite it with the local version

That's what I think should happen, however haven't really tested this path through install command - would have to take a look whether the behavior is indeed different.

Is there any specific reason for using local file in conjunction with install command?

**Comment by 0x53A** [comment]
I may be doing something unneccessary complicated, I'm not that familiar with paket, I use it for a few toy projects.

Let's say I have configured a paket.local to override one package with a local version.

Then I edit one of the paket.dependencies, because one project has a new depedendency.
Then if I don't misunderstand paket completely, I need to run paket install, to add the assemblies refs to the csproj/fsproj.

My understanding of the current behaviour is, that the install would then replace the local version from paket.local with the server version?
So I would need to run a paket restore afterwards to get the local version again?

**Comment by theimowski** [comment]
> My understanding of the current behaviour is, that the install would then replace the local version from paket.local with the server version?
So I would need to run a paket restore afterwards to get the local version again?

That's something I'd have to check - not sure what's the current behavior.

My question regarding running install with overrides present in paket.local was due to I treat the overrides as "temporary", i.e. for the very purpose of testing a local package. 
If I need to run install command, I'd rather finished the "local" mode first by commenting / removing overrides, and only then run install command.
But there might be some use case I'm missing, hence asking

**Comment by theimowski** [comment]
I can confirm that current behavior is that local file works only with the restore command, and looks like adding it to the install command would require a bit more effort

**Comment by daz10000** [comment]
I don't know if this is related (apologies if I'm hijacking thread) but I can't for the life of me get paket to use my local reference.   I have an updated version of a dependent library local (and I bumped the version number).  As the thread indicates (thanks!) restore is the only operation that seems to observe the paket.local (that wasn't obvious in the docs, in fact I'm using project scaffold and was naively hoping that ./build.sh would do the right thing).    

Running paket restore indicates that it's using the local path to the updated dependent library but it doesn't actually seem to copy anything.  Maybe because the updated library has a higher version number?   It burned me a few times before I realized I wasn't actually testing the updated library when I tested my project.  I copied it over manually and did some testing but running projectscaffold build stubbornly reverts my manual copy (presumably paket happily restoring it to the published version rather than my paket.local dependency).  I've tried everything I can think of 
```
$ .paket/paket.exe restore
Paket version 5.2.1
paket.local override: nuget GslCore group Main -> c:\seq\GslCore\bin
Performance:
 - Disk IO: 221 milliseconds
 - Runtime: 1 second
```

If I run that same operation verbose, it mentions obviously the same override line but in addition here are the library specific operations - and it doesn't seem to touch the c:\seq\GslCore override folder ever again.  I'm completely puzzled.

```
GslCore 0.4.21 unzipped to C:\seq\Gsl\packages\GslCore
libraries found in C:\seq\Gsl\packages\GslCore matching *.*:
  - C:\seq\Gsl\packages\GslCore\lib\net461\GslCore.dll
No libraries found in C:\seq\Gsl\packages\GslCore matching *.*
No libraries found in C:\seq\Gsl\packages\GslCore matching *.*
No .targets files found in C:\seq\Gsl\packages\GslCore matching *.*
No analyzer dlls found in C:\seq\Gsl\packages\GslCore matching *.dll
```

**Comment by daz10000** [comment]
Oh... just figured out why it's doing that.   The bumped version number is the issue.   If I omit the updated release_notes.md info and build it exactly as though the new version has the same version number, then paket seems to copy it over.  I can sort of see why you'd want to do that to avoid having to temporarily update the paket.dependencies info, but it's really counter-intuitive and could use some documentation :(  )  I've lost quite a few hours troubleshooting this.

It's also slightly counter to actually testing the final product.   If project B depends on project A and I make some changes to A and want to realistically test the whole setup before pushing,   I'd like to first bump A's version number,  then update B's dependencies to the new A version, insert paket.local to provide the new version of A and then test it all exactly as it's going to work.  If the testing goes well,  I remove the paket.local and push the projects A and B out (or just A and test that I can build B from the pushed packages).  Pretending the new A is the same as the old version while doing testing seems dangerous...

**Comment by theimowski** [comment]
If you want to use a different version you need to specify it in paket.local:

```
nuget GslCore group Main -> c:\seq\GslCore\bin version 2.8.8
```

this is documented [here](https://fsprojects.github.io/Paket/local-file.html#Nupkg-version) (Nupkg version)

> Pretending the new A is the same as the old version while doing testing seems dangerous...

Agreed, that's why we issue a warning so that one knows what he's doing. All the stuff related to paket.local and overriding is really somewhat not in spirit of Paket main goal i.e. lock file. Because of this we want to make sure the overrides do not leak outside local environment

**Comment by daz10000** [comment]
Thanks guys for the timely and thoughtful feedback and for the great tool.   I reviewed the docs and the various places I went wrong (also had chance to watch a colleague work through same process and make some similar errors today), so I think my final feedback would still be not necessarily to change the tool at all but maybe reinforce the docs.    The wrong path I (and others) seem to have taken is to either try to 
1) upgrade package version locally and miss that paket isn't picking up the new version ( there doesn't seem to be any warning associated with this misunderstanding and it quietly tests against the public version which can be deadly for testing)

2) upgrade package and try to use paket install  / update  (this does give an error but it's confusing unless you know that only restore observes the paket.local directive).   I and a colleague had even indepdently resorted to hacking the lock file to get around this which is not ideal user behavior.

3) miss the option of specifying version.  I made a further mistake here assuming that the version >= myNewVersion   syntax would work and got a further failure mode

Maybe a prominent warning early in the paket.lock example that it only affects restore and by default can only substitute a file with the exact version as the current paket.lock entry would suffice (with a pointer to the version section at the end if you need to change that).  I get that this is off the beaten path for paket but as someone who has to release two part systems with dependencies,  a simple safe way to substitute the about to be released library is very helpful.  I won't make these mistakes again but I tried pretty hard to follow docs (even did a site: paket.local search against the whole site to see what I missing),  and got burned a little so some doc tweaks might save the next guy  (and thanks for an awesome tool, really!)

**Comment by forki** [comment]
We would love to get a pull request to the docs which would make it easier
to you and others. Please send one. It would be very welcome

Am 27.06.2017 02:37 schrieb "daz10000" <notifications@github.com>:

**Comment by daz10000** [comment]
Have sent you a pull request (this is my first for a public repo so be gentle!)

**Comment by enricosada** [comment]
Closing, by design
